moshi-sealed/metadata-reflect/src/main/kotlin/dev/zacsweers/moshix/sealed/reflect/MetadataMoshiSealedJsonAdapterFactory.kt

```diff
@@ -0,0 +1,152 @@
+package dev.zacsweers.moshix.sealed.reflect
+
+import com.squareup.moshi.JsonAdapter
+import com.squareup.moshi.JsonClass
+import com.squareup.moshi.Moshi
+import com.squareup.moshi.Types
+import com.squareup.moshi.adapters.PolymorphicJsonAdapterFactory
+import dev.zacsweers.moshix.sealed.annotations.DefaultNull
+import dev.zacsweers.moshix.sealed.annotations.DefaultObject
+import dev.zacsweers.moshix.sealed.annotations.TypeLabel
+import dev.zacsweers.moshix.sealed.runtime.internal.ObjectJsonAdapter
+import kotlinx.metadata.ClassName
+import kotlinx.metadata.Flag
+import kotlinx.metadata.KmClass
+import kotlinx.metadata.jvm.KotlinClassHeader
+import kotlinx.metadata.jvm.KotlinClassMetadata
+import java.lang.reflect.Type
+
+/** Classes annotated with this are eligible for this adapter. */
+private val KOTLIN_METADATA = Metadata::class.java
+
+private val UNSET = Any()
+
+public class MetadataMoshiSealedJsonAdapterFactory : JsonAdapter.Factory {
+  override fun create(type: Type,
+                      annotations: MutableSet<out Annotation>,
+                      moshi: Moshi): JsonAdapter<*>? {
+    if (annotations.isNotEmpty()) {
+      return null
+    }
+    val rawType = Types.getRawType(type)
+    if (!rawType.isAnnotationPresent(KOTLIN_METADATA)) return null
+    val kmClass = rawType.header()?.toKmClass() ?: return null
```

Comment:
To optimize performance a bit, lets defer this creation to _after_ we've checked for `@JsonClass` and the `sealed:` prefix